src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -582,162 +1026,155 @@ private void onMqttUnsubscribe(
                 topics++;
             }
 
-            doMqttUnsuback(topics);
+            doEncodeUnsuback(traceId, authorization, topics);
         }
 
-        private void onMqttPublish(
+        private void onDecodePublish(
+            long authorization,
+            MqttPublishStream publishStream,
+            String topicName,
+            long newRouteId,
+            long newInitialId,
+            long newReplyId,
+            MessageConsumer newTarget,
+            MqttDataExFW dataEx,
+            OctetsFW payload,
             MqttPublishFW publish)
         {
-            final String topicName = publish.topicName().asString();
-
-            String info = "info";
-
-            final MqttDataExFW dataEx = mqttDataExRW
-                .wrap(extBuffer, 0, extBuffer.capacity())
-                .typeId(mqttTypeId)
-                .topic(topicName)
-                .expiryInterval(15)
-                .contentType("message")
-                .format(f -> f.set(MqttPayloadFormat.TEXT))
-                .responseTopic(topicName)
-                .correlationInfo(c -> c.bytes(b -> b.set(info.getBytes(UTF_8))))
-                .build();
-
-            OctetsFW payload = publish.payload();
-
-            MqttPublishStream publishStream = publishers.get(topicName);
-
-            final RouteFW route = resolveTarget(routeId, authorization, topicName);
-            if (route != null)
+            if (publishStream != null)
```

Comment:
`MqttPublishStream` should ideally only be used inside `MqttServer` not in shared `decodePublish` method on factory.